### PR TITLE
Implement user agent parsing with exact aws behavior

### LINKED ISSUE #556: Complete user agent detection configuration for YAUAA
open todo from https://github.com/mozilla/gcp-ingestion/blob/master/ingestion-beam/src/main/resources/UserAgents/Firefox.yaml#L8

### PR DESCRIPTION
**Pull request by relud**
fixes #556

### CODE REVIEW

File: ingestion-beam/src/test/java/com/mozilla/telemetry/decoder/ParseUserAgentTest.java [hunk 1]
```diff
@@ -33,7 +33,7 @@ public void testOutput() {
         "{\"attributeMap\":{\"host\":\"test\"},\"payload\":\"dGVzdA==\"}",
         "{\"attributeMap\":{},\"payload\":\"dGVzdA==\"}",
         "{\"attributeMap\":" + "{\"user_agent_browser\":\"Firefox\""
-            + ",\"user_agent_version\":\"63.0\"" + ",\"user_agent_os\":\"Macintosh\""
+            + ",\"user_agent_version\":\"63\"" + ",\"user_agent_os\":\"Macintosh\""
             + "},\"payload\":\"\"}");
```
Comment (relud) [inline ingestion-beam/src/test/java/com/mozilla/telemetry/decoder/ParseUserAgentTest.java]: needs more tests

**Comment by relud** [comment]
@jklukas I tried to remove the yauaa dependency, but we import
`nl.basjes.shaded.org.springframework.core.io.Resource` and
`nl.basjes.shaded.org.springframework.core.io.support.PathMatchingResourcePatternResolver`. I tried replacing yauaa with spring-core `5.1.9.RELEASE`, but that didn't fix the issue and I'm not sure what version yauaa is shading.

File: ingestion-beam/src/test/java/com/mozilla/telemetry/decoder/ParseUserAgentTest.java [hunk 2]
```diff
@@ -20,21 +25,69 @@
...
-            + "},\"payload\":\"\"}");
+  private static final String BROWSER = "user_agent_browser";
+  private static final String OS = "user_agent_os";
+  private static final String VERSION = "user_agent_version";
+
+  private static final Map<String, Map<String, String>> osSingleMatch = new ImmutableMap //
+      .Builder<String, Map<String, String>>()
+          .put("Firefox AndroidSync", ImmutableMap.of(BROWSER, "FxSync", OS, "Android"))
+          .put("Firefox-iOS-Sync", ImmutableMap.of(BROWSER, "FxSync", OS, "iOS"))
+          .put("iPod", ImmutableMap.of(OS, "iPod")).put("iPad", ImmutableMap.of(OS, "iPad"))
```
Comment (jklukas) [inline ingestion-beam/src/test/java/com/mozilla/telemetry/decoder/ParseUserAgentTest.java]:
```suggestion
          .put("iPod", ImmutableMap.of(OS, "iPod")) //
          .put("iPad", ImmutableMap.of(OS, "iPad")) //
```

File: ingestion-beam/src/test/java/com/mozilla/telemetry/decoder/ParseUserAgentTest.java [hunk 3]
```diff
@@ -20,21 +25,69 @@
...
+          .put("BlackBerry", ImmutableMap.of(OS, "BlackBerry"))
+          .put("Macintosh", ImmutableMap.of(OS, "Macintosh"))
+          .put("Mozilla/5.0 (Mobile;", ImmutableMap.of(OS, "FirefoxOS"))
+          .put("Windows NT 10.0", ImmutableMap.of(OS, "Windows 10"))
+          .put("Windows NT 6.3", ImmutableMap.of(OS, "Windows 8.1"))
+          .put("Windows NT 6.2", ImmutableMap.of(OS, "Windows 8"))
+          .put("Windows NT 6.1", ImmutableMap.of(OS, "Windows 7"))
+          .put("Windows NT 6.0", ImmutableMap.of(OS, "Windows Vista"))
+          .put("Windows NT 5.1", ImmutableMap.of(OS, "Windows XP"))
+          .put("Windows NT 5.0", ImmutableMap.of(OS, "Windows 2000")).build();
```
Comment (jklukas) [inline ingestion-beam/src/test/java/com/mozilla/telemetry/decoder/ParseUserAgentTest.java]:
```suggestion
          .put("Windows NT 5.0", ImmutableMap.of(OS, "Windows 2000")) //
          .build();
```

File: ingestion-beam/src/test/java/com/mozilla/telemetry/decoder/ParseUserAgentTest.java [hunk 4]
```diff
@@ -20,21 +25,69 @@
...
+          .put("MSIE", ImmutableMap.of(BROWSER, "MSIE"))
+          .put("Trident/7.0", ImmutableMap.of(BROWSER, "MSIE", VERSION, "11"))
+          .put("Safari", ImmutableMap.of(BROWSER, "Safari"))
+          .put("Firefox AndroidSync 4",
+              ImmutableMap.of(BROWSER, "FxSync", VERSION, "4", OS, "Android"))
+          .put("Firefox AndroidSync", ImmutableMap.of(BROWSER, "FxSync", OS, "Android"))
+          .put("Firefox-iOS-Sync/5", ImmutableMap.of(BROWSER, "FxSync", VERSION, "5", OS, "iOS"))
+          .put("Firefox-iOS-Sync", ImmutableMap.of(BROWSER, "FxSync", OS, "iOS"))
+          .put("Firefox/6", ImmutableMap.of(BROWSER, "Firefox", VERSION, "6"))
+          .put("Firefox", ImmutableMap.of(BROWSER, "Firefox")).build();
```
Comment (jklukas) [inline ingestion-beam/src/test/java/com/mozilla/telemetry/decoder/ParseUserAgentTest.java]:
```suggestion
          .put("Firefox", ImmutableMap.of(BROWSER, "Firefox")) //
          .build();
```

File: ingestion-beam/src/test/java/com/mozilla/telemetry/decoder/ParseUserAgentTest.java [hunk 5]
```diff
@@ -47,4 +100,67 @@ public void testOutput() {
...
+    test(builder.build());
+  }
+
+  @Test
+  public void testBrowserSingleMatch() {
+    test(browserSingleMatch);
+  }
+
+  @Test
+  public void testBrowserBestMatcFirst() {
```
Comment (jklukas) [inline ingestion-beam/src/test/java/com/mozilla/telemetry/decoder/ParseUserAgentTest.java]:
```suggestion
  public void testBrowserBestMatchFirst() {
```

**Comment by jklukas** [comment]
> I tried replacing yauaa with spring-core 5.1.9.RELEASE, but that didn't fix the issue and I'm not sure what version yauaa is shading.

I can take a look as a follow-up after this is merged.

**Pull request merged**